### ISSUE TITLE
Enforce consistency with variant naming

### ISSUE THREAD

**Issue by sethrj** [spack/spack#15239]
I was surprised by a new variant name in OpenBLAS being mixed case (https://github.com/spack/spack/pull/14876), since I thought variants were strictly `[a-z0-9-]+`. A grep of `builtin/packages/*/package.py` shows a number of inconsistencies:
- That new `consistentFPCSR` variant is the only variant in any package to have mixed case
- The only other variants to have upper case are `X`, `X11`, and a variant `K` in the `eem` package
- `X`, `X11`, `x`, and `x11`  are all variants meaning "require the `libx11` package"
- Most variants use underscores for separators if at all, but some use hyphens
- There are surely gobs more.

Questions:
1. Is there a point at which we unify these variant names?
2. Is there documentation for best practices? (see https://github.com/spack/spack/issues/557 but a quick check in the [packaging guide](https://spack.readthedocs.io/en/latest/packaging_guide.html#variants) and [packaging creation tutorial](https://spack-tutorial.readthedocs.io/en/latest/tutorial_packaging.html#variants) revealed nothing)
3. How can we prevent new odd cases from creeping in during MRs?

**Comment by adamjstewart** [comment]
In `lib/spack/spack/cmd/common.py`, we have common command-line flags and arguments shared across Spack commands. I wonder if we could do something similar for variants. We started down this direction by adding common variants for build systems, like the `build_type` variant in `CMakePackage`.

As for mix-cased and dashes vs. underscores, that should be much easier to enforce. We could restrict the possible set of characters allowed in variant names. Personally I'm for lowercase and underscores since `-` is already used for variant disabling and package names can also contain dashes.

**Comment by sethrj** [comment]
Agreed. Plus `build_type` isn't going to change to `build-type` anytime soon 😉

**Comment by glennpj** [comment]
> Personally I'm for lowercase and underscores since `-` is already used for variant disabling and package names can also contain dashes.

Yes, but, configuration options can use dashes so if one wishes to use `with_or_without` or `enable_or_disable` then `-` needs to be allowed in the variant name. Also, the spack documentation says that to use `-` to disable a variant then it needs a space in front of it.

**Comment by adamjstewart** [comment]
> Yes, but, configuration options can use dashes so if one wishes to use with_or_without or enable_or_disable then - needs to be allowed in the variant name.

We could tell these functions to automatically replace `_` with `-`. Not perfect, but these functions also don't need to be used in the rare case of a configure flag that uses underscores.

**Comment by sethrj** [comment]
Aesthetically I think dashes look better (and generally command-line interfaces are tending toward dashes rather than underscores to save your `shift` key some work) but it would look almost indistinguishable from `~` when spack prints specs:
```
veccore@0.6.1b1 build_type=RelWithDebInfo ~cuda cuda_arch=none
```
vs
```
veccore@0.6.1b1 build-type=RelWithDebInfo ~cuda cuda-arch=none
```
and even worse an imaginary library
```
py-something@1.0.0~bar~foo+foo-bar
```

**Comment by sethrj** [comment]
Honestly I feel like both dashes and underscores should be forbidden or at least strongly discouraged. The common `cxxstd` variant could easily have been written `cxx-std`, but it's not. `build_type` is still arguably readable as `buildtype` or (since CMake also refers to "`CMAKE_C_FLAGS_<CONFIG>`") perhaps just `config` to get around the issue of using a phrase rather than a word as a variant.

**Comment by adrienbernede** [comment]
@sethrj I am running into an issue that lead me to this thread.

Interestingly, the following is not a valid spec, because `~cuda` is preceeded by a whitespace. 
`veccore@0.6.1b1 build_type=RelWithDebInfo ~cuda cuda_arch=none`

The correct spec would be:
`veccore@0.6.1b1 build_type=RelWithDebInfo -cuda cuda_arch=none`

However spack would generate the former, instead of the latter.

That is to say that it’s likely that Spack should generate the latter, which you may argue is even more confusing. Now imagine if the variant was (ill-)named `with-cuda`...

The naming confusion is not a priority to me though. I just found this worth noting.
